src/supplemental_attribute_associations.jl

```diff
@@ -0,0 +1,401 @@
+    return Base.UUID.(table.component_uuid)
+end
+
+"""
+Return the supplemental attribute UUIDs associated with the component and attribute type.
+"""
+function list_associated_supplemental_attribute_uuids(
+    associations::SupplementalAttributeAssociations,
+    component::InfrastructureSystemsComponent;
+    attribute_type::Union{Nothing, Type{<:SupplementalAttribute}} = nothing,
+)
+    c_str = "component_uuid = '$(get_uuid(component))'"
+    where_clause = if isnothing(attribute_type)
+        c_str
+    else
+        a_str = _get_attribute_type_string(attribute_type)
+        "$a_str AND $c_str"
+    end
+    query = """
+        SELECT attribute_uuid
+        FROM $SUPPLEMENTAL_ATTRIBUTE_TABLE_NAME
+        WHERE $where_clause
+    """
+    table = Tables.columntable(_execute(associations, query))
+    return Base.UUID.(table.attribute_uuid)
+end
+
+"""
+Remove the association between the attribute and component.
+"""
+function remove_association!(
+    associations::SupplementalAttributeAssociations,
+    component::InfrastructureSystemsComponent,
+    attribute::SupplementalAttribute,
+)
+    a_uuid = get_uuid(attribute)
+    c_uuid = get_uuid(component)
+    where_clause = "WHERE attribute_uuid = '$a_uuid' AND component_uuid = '$c_uuid'"
+    num_deleted = _remove_associations!(associations, where_clause)
+    if num_deleted != 1
+        error("Bug: unexpected number of deletions: $num_deleted. Should have been 1.")
+    end
+end
+
+"""
+Remove all associations of the given type.
+"""
+function remove_associations!(
+    associations::SupplementalAttributeAssociations,
+    type::Type{<:SupplementalAttribute},
+)
+    where_clause = "WHERE attribute_type = '$(nameof(type))'"
+    num_deleted = _remove_associations!(associations, where_clause)
+    @debug "Deleted $num_deleted supplemental attribute associations" _group =
+        LOG_GROUP_SUPPLEMENTAL_ATTRIBUTES
+    return
+end
+
+"""
+Replace the component UUID in the table.
+"""
+function replace_component_uuid!(
+    associations::SupplementalAttributeAssociations,
+    old_uuid::Base.UUID,
+    new_uuid::Base.UUID,
+)
+    query = """
+        UPDATE $SUPPLEMENTAL_ATTRIBUTE_TABLE_NAME
+        SET component_uuid = '$new_uuid'
+        WHERE component_uuid = '$old_uuid'
+    """
+    _execute(associations, query)
+    return
+end
+
+"""
+Run a query and return the results in a DataFrame.
+"""
+function sql(associations::SupplementalAttributeAssociations, query::String)
+    return DataFrames.DataFrame(_execute(associations, query))
+end
+
+"""
+Return all rows in the table as dictionaries.
+"""
+function to_records(associations::SupplementalAttributeAssociations)
+    query = "SELECT * FROM $SUPPLEMENTAL_ATTRIBUTE_TABLE_NAME"
+    return Tables.rowtable(_execute(associations, query))
+end
+
+"""
+Add records to the database. Expects output from [`to_records`](@ref).
+"""
+function load_records!(associations::SupplementalAttributeAssociations, records)
```

Comment:
What's a record in this function?

Reply:
Rows from the database. When we serialize the system to JSON we write all database rows (`to_records`) to the JSON file. When we de-serialize from JSON we load those records back into the database.